File: modules/api/src/main/scala/org/bigasterisk/api/Watchpoint.scala
package org.bigasterisk.api

import org.apache.spark.sql.{Column, DataFrame, Row}

/**
 * A guard placed on intermediate data: rows flowing through a point in a query are
 * tested against a condition, and the ones that match are counted and sampled back to
 * the driver.
 *
 * This is BigDebug's on-demand watchpoint (ICSE 2016) — the distributed counterpart of
 * a watchpoint in a conventional debugger. Instead of inspecting a variable at a line
 * of code, you inspect the records passing a stage of a distributed job, without
 * collecting the whole intermediate dataset.
 *
 * @group watchpoint
 */
trait Watchpoint {

  /** Identifier, unique within the session. Appears in the Spark UI as the accumulator name. */
  def id: String

  /** The guard, as SQL text. */
  def condition: String

  /**
   * The instrumented DataFrame. Build the rest of the query on this rather than on the
   * DataFrame you passed to `watch`, or nothing will be observed.
   */
  def df: DataFrame

  /**
   * How many rows matched the condition.
   *
   * Counted in full even when more rows matched than [[capacity]] allows to be kept,
   * so this is the true selectivity of the guard rather than a sample size.
   */
  def hits: Long

  /** The matching rows retained, up to [[capacity]]. */
  def captured: Array[Row]

  /**
   * The same rows as [[captured]], each serialised as a JSON object.
   *
   * Provided for language bindings that cannot marshal a Spark `Row` across the
   * process boundary — the PySpark front end reads this.
   */
  def capturedJson: Array[String]

  /** The most rows this watchpoint will bring back to the driver. */
  def capacity: Int

  /** True when more rows matched than were kept — [[captured]] is a sample. */
  final def truncated: Boolean = hits > captured.length

  /** Discards what has been observed so far, so the watchpoint can be reused for another run. */
  def reset(): Unit
}

/**
 * On-demand watchpoints over the intermediate data of a Spark SQL query.
 *
 * Obtain an instance from [[BigAsterisk.watchpoints]].
 *
 * {{{
 * val orders = spark.table("orders")
 * val wp = BigAsterisk.watchpoints(spark).watch(orders, col("amount") > 10000)
 *
 * // build the rest of the query on the instrumented DataFrame
 * wp.df.groupBy("cid").sum("amount").collect()
 *
 * println(s"\${wp.hits} suspicious rows")
 * wp.captured.foreach(println)
 * }}}
 *
 * @group watchpoint
 */
trait WatchpointSupport {

  /**
   * Places a watchpoint on the rows flowing out of `df`.
   *
   * Evaluation happens on the executors and is fused into Spark's generated code, so a
   * watchpoint costs one predicate per row and moves no data until something matches.
   * Only the first `capacity` matching rows are brought back; the rest are counted.
   *
   * @param capacity the most matching rows to retain. Keep it small: these rows are
   *        held on the driver.
   */
  def watch(df: DataFrame, condition: Column, capacity: Int = 1000): Watchpoint

  /** Every watchpoint created in this session that has not been cleared. */
  def active: Seq[Watchpoint]

  /** Forgets every watchpoint. Already-instrumented DataFrames keep working. */
  def clear(): Unit
}

File: python/bigasterisk/watchpoint.py
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


"""On-demand watchpoints over the intermediate data of a Spark SQL query.

A watchpoint is a guard on records flowing through a point in a query: matching rows
are counted and sampled back to the driver, without collecting the intermediate
dataset. This is BigDebug's watchpoint primitive (ICSE 2016).

    import bigasterisk
    from pyspark.sql.functions import col

    orders = spark.table("orders")
    wp = bigasterisk.watchpoints(spark).watch(orders, col("amount") > 10000)

    # build the rest of the query on the instrumented DataFrame
    wp.df.groupBy("cid").sum("amount").collect()

    print("%d suspicious rows" % wp.hits)
    for row in wp.captured:
        print(row)
"""

import json


def _wrap_dataframe(jdf, spark):
    """Wrap a JVM DataFrame as a PySpark DataFrame.

    PySpark has moved this constructor between releases, so try the session-based form
    first and fall back to the older SQLContext form.
    """
    from pyspark.sql import DataFrame

    try:
        return DataFrame(jdf, spark)
    except TypeError:
        return DataFrame(jdf, spark._wrapped)


class Watchpoint:
    """A guard on the rows flowing out of a DataFrame."""

    def __init__(self, jwatchpoint, spark):
        self._j = jwatchpoint
        self._spark = spark

    @property
    def id(self):
        """Identifier, unique in the session. Names the accumulator in the Spark UI."""
        return self._j.id()

    @property
    def condition(self):
        """The guard, as SQL text."""
        return self._j.condition()

    @property
    def df(self):
        """The instrumented DataFrame.

        Build the rest of your query on this, not on the DataFrame you passed to
        :meth:`Watchpoints.watch`, or nothing will be observed.
        """
        return _wrap_dataframe(self._j.df(), self._spark)

    @property
    def hits(self):
        """How many rows matched — counted in full, even beyond ``capacity``."""
        return self._j.hits()

    @property
    def capacity(self):
        """The most matching rows this watchpoint will bring back to the driver."""
        return self._j.capacity()

    @property
    def captured(self):
        """The matching rows retained, as a list of dicts."""
        return [json.loads(s) for s in self._j.capturedJson()]

    @property
    def truncated(self):
        """True when more rows matched than were kept, so ``captured`` is a sample."""
        return self._j.truncated()

    def reset(self):
        """Discard what has been observed, so the watchpoint can be reused for a new run."""
        self._j.reset()

    def __repr__(self):
        return "Watchpoint(%s, condition=%s, hits=%d%s)" % (
            self.id, self.condition, self.hits, ", truncated" if self.truncated else "")


class Watchpoints:
    """Watchpoint creation and registry for a SparkSession."""

    def __init__(self, spark):
        self._spark = spark
        self._support = spark._jvm.org.bigasterisk.api.BigAsterisk.watchpoints(
            spark._jsparkSession)

    def watch(self, df, condition, capacity=1000):
        """Place a watchpoint on the rows flowing out of ``df``.

        ``condition`` is a PySpark ``Column``. Evaluation happens on the executors and
        is fused into Spark's generated code, so an unmatched row costs one predicate
        and moves no data.
        """
        return Watchpoint(
            self._support.watch(df._jdf, condition._jc, int(capacity)), self._spark)

    def active(self):
        """Every watchpoint created and not yet cleared."""
        j = self._support.active()
        return [Watchpoint(j.apply(i), self._spark) for i in range(j.size())]

    def clear(self):
        """Forget every watchpoint. Already-instrumented DataFrames keep working."""
        self._support.clear()
